OrbitQt/ConnectToStadiaWidget.cpp

```diff
@@ -0,0 +1,344 @@
+
+#include "ConnectToStadiaWidget.h"
+
+#include <QMessageBox>
+#include <QRadioButton>
+#include <QSettings>
+#include <optional>
+
+#include "Error.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/Result.h"
+#include "OrbitGgp/Client.h"
+#include "OverlayWidget.h"
+#include "servicedeploymanager.h"
+
+namespace {
+const QString kRememberChosenInstance{"RememberChosenInstance"};
+}  // namespace
+
+namespace orbit_qt {
+
+using OrbitGgp::Instance;
+using OrbitSshQt::ScopedConnection;
+
+ConnectToStadiaWidget::ConnectToStadiaWidget(QWidget* parent)
+    : QWidget(parent),
+      ui_(std::make_unique<Ui::ConnectToStadiaWidget>()),
+      settings_("The Orbit Authors", "Orbit Profiler") {
+  ui_->setupUi(this);
+
+  if (!settings_.value(kRememberChosenInstance).toString().isEmpty()) {
+    remembered_instance_id_ = settings_.value(kRememberChosenInstance).toString();
+    ui_->rememberCheckBox->setChecked(true);
+  }
+
+  ui_->instancesTableView->setModel(&instance_model_);
+
+  QObject::connect(ui_->connectToStadiaInstanceRadioButton, &QRadioButton::toggled, this,
+                   [this](bool checked) {
+                     if (checked) {
+                       emit Selected();
+                     } else {
+                       ui_->connectToStadiaInstanceRadioButton->setChecked(true);
+                     }
+                   });
+
+  QObject::connect(this, &ConnectToStadiaWidget::ErrorOccurred, this,
+                   [this](const QString& message) {
+                     QMessageBox::critical(this, QApplication::applicationName(), message);
+                   });
+
+  QObject::connect(ui_->instancesTableView->selectionModel(), &QItemSelectionModel::currentChanged,
+                   this, [this](const QModelIndex& current) {
+                     if (!current.isValid()) return;
+
+                     CHECK(current.model() == &instance_model_);
+                     connection_artifacts_->selected_instance_ =
+                         current.data(Qt::UserRole).value<Instance>();
+                     emit InstanceSelected();
+                   });
+
+  QObject::connect(ui_->rememberCheckBox, &QCheckBox::toggled, this, [this](bool checked) {
+    if (checked) {
+      CHECK(connection_artifacts_->selected_instance_ != std::nullopt);
+      settings_.setValue(kRememberChosenInstance, connection_artifacts_->selected_instance_->id);
+    } else {
+      settings_.remove(kRememberChosenInstance);
+    }
+  });
+
+  SetupAndStartStateMachine();
+}
+
+void ConnectToStadiaWidget::SetConnectionArtifacts(ConnectionArtifacts* connection_artifacts) {
+  CHECK(connection_artifacts != nullptr);
+  connection_artifacts_ = connection_artifacts;
+}
+
+void ConnectToStadiaWidget::SetupAndStartStateMachine() {
+  state_machine_.setGlobalRestorePolicy(QStateMachine::RestoreProperties);
+
+  // STATES list
+  auto* s_ggp_not_available = new QState();
+  state_machine_.addState(s_ggp_not_available);
+  auto* s_instances_empty = new QState();
```

Comment:
Is `s_instances_empty` a state that is actually needed? Isn't this just a special case of `s_instances_loaded` with 0 instances? As far as I see, the state of `s_instances_loaded` without an instance selected is equal to `s_instances_empty`?

Reply:
You are right. I merged both states into `s_idle`. Thanks for finding that!